database/mongo/controller.mjs

```diff
@@ -0,0 +1,237 @@
+            const id = this.newID()
+            data["_id"] = id
+            const result = await this.db.collection(collection).insertOne(data)
+            if (result.insertedId) {
+                return data
+            } else {
+                return { "endpoint_error": "insertOne", "status": 500, "message": "Document was not inserted into the database." }
+            }
+        } catch (err) {
+            console.error(err)
+            return { "endpoint_error": "insertOne", "status": 500, "message": "There was an error inserting the document into the database." }
+        }
+    }
+
+    /**
+     * Update an existing object in the database (mongo)
+     * @param data JSON from an HTTP POST request.  It must contain an id.
+     * @return The inserted document JSON or error JSON
+     */
+    async update(data) {
+        try {
+            //need to determine what collection (projects, groups, userPerferences) this goes into.
+            const data_type = data["@type"] ?? data.type ?? null
+            let data_id = data["@id"] ?? data._id ?? null
+            let collection = null
+            if (!data_id)
+                return { "endpoint_error": "updateOne", "status": 400, "message": `Cannot find 'type' on this data, and so cannot figure out a collection for it.` }
+            if (!data_type)
+                return { "endpoint_error": "updateOne", "status": 400, "message": `Cannot find 'type' on this data, and so cannot figure out a collection for it.` }
+            collection = discernCollectionFromType(data_type)
+            if (!collection)
+                return { "endpoint_error": "updateOne", "status": 500, "message": `Cannot figure which collection for object of type '${data_type}'` }
+            const obj_id = data_id.split("/").pop()
+            const filter = { "_id": data_id }
+            const result = await this.db.collection(collection).replaceOne(filter, data)
+            if (result?.matchedCount === 0) {
+                return { "endpoint_error": "updateOne", "status": 404, "message": `id '${obj_id}' Not Found` }
+            }
+            if (result?.modifiedCount >= 0) {
+                return data
+            } else {
+                return { "endpoint_error": "updateOne", "status": 500, "message": "Document was not updated in the database." }
+            }
+        } catch (err) {
+            console.error(err)
+            return { "endpoint_error": "updateOne", "status": 500, "message": "There was an error updating the document in the database." }
+        }
+    }
+
+    /**
+     * Update an existing object in the database (mongo)
+     * @param data JSON from an HTTP DELETE request.  It must contain an id.
+     * @return The delete result JSON or error JSON
+     */
+    async remove(id) {
+        return { "endpoint_error": "deleteOne", "status": 501, "message": `Not yet implemented.  Stay tuned.` }
```

Comment:
(aside) these are bad examples for the OSS folks because they are "error-like objects" without actually being Errors.